README.md
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/d8e41d16dc114f95806e050515c2dbc0)](https://www.codacy.com/gh/zie87/wth_woth/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=zie87/wth_woth&amp;utm_campaign=Badge_Grade)
[![Coverage Status](https://coveralls.io/repos/github/zie87/wth_woth/badge.svg)](https://coveralls.io/github/zie87/wth_woth)

# Wagic: War of the Homebrew

*War of the Homebrew* is an open source collectible trading card game. You play as a wizard against the CPU with your army of minions. You start the game with a small collection of cards (symbolizing creatures and spells), and as you win battles against the AI, you earn credits that allow you to buy more cards and build more powerful decks. As you play, you unlock cards and new game modes, so it is much more than a simple card simulation. 

*WotH* features several thousands of cards (*currently over 9000 unique cards*) and is playable on multiple platforms.

![screenshot](./doc/screenshot_demo.png)

It is based on [Wagic the Homebrew] in version 0.18.6 ([released](https://wololo.net/download/) in 2012). *War of the Homebrew* should be compatible with all mods create for this particular *Wagic* version.

[Wagic the Homebrew]: https://github.com/WagicProject/wagic

## Documentation

* [Manual](./doc/manual/manual.md)
* [Building](./doc/build.md)
* [Scripting](./doc/card_scripting.md)


doc/history.md
# WotH 0.0.0

| Language         |    Files    |    Lines     |   Blank    |  Comment    |     Code |
|:-----------------|------------:|-------------:|-----------:|------------:|---------:|
| C/C++ Header     |     3179    |   519960     |   72888    |   110806    |   336266 |
| C                |      289    |   163656     |   19046    |    19668    |   124942 |
| C++              |      186    |    94334     |   11888    |     5079    |    77367 |
| Plain Text       |      922    |    51202     |     472    |        0    |    50730 |
| Objective-C      |      127    |    36724     |    6482    |     4185    |    26057 |
| HTML             |       93    |    23829     |     887    |       91    |    22851 |
| Perl             |        8    |     1733     |     218    |      124    |     1391 |
| Java             |        2    |     1342     |     221    |      172    |      949 |
| Prolog           |        2    |      931     |      56    |       26    |      849 |
| Makefile         |       14    |      737     |     131    |       26    |      580 |
| XML              |        9    |      606     |      71    |       59    |      476 |
| CSS              |        2    |      460     |      20    |        6    |      434 |
| Python           |        4    |      518     |      98    |       83    |      337 |
| Objective-C++    |        1    |      405     |      47    |       22    |      336 |
| Qml              |        2    |      224     |      26    |        9    |      189 |
| Assembly         |        1    |      214     |      20    |       84    |      110 |
| Autoconf         |        2    |      297     |      26    |      227    |       44 |
| Markdown         |        2    |       44     |      14    |        0    |       30 |
| Batch            |        4    |       31     |       1    |        0    |       30 |
| Ruby             |        1    |       11     |       1    |        1    |        9 |
|                  |             |              |            |             |          |
| **Total**        |     4850    |   897258     |  112613    |   140668    |   643977 |



# WotH 0.0.99 [remove qt4]


| Language         |    Files    |    Lines     |   Blank    |  Comment    |     Code |
|:-----------------|------------:|-------------:|-----------:|------------:|---------:|
| C++              |      179    |    73736     |    9906    |     4524    |    59306 |
| Plain Text       |      926    |    51491     |     503    |        0    |    50988 |
| C/C++ Header     |      209    |    30826     |    4604    |     6019    |    20203 |
| HTML             |        2    |     9154     |       6    |        0    |     9148 |
| C                |        9    |     3107     |     306    |      206    |     2595 |
| Markdown         |        5    |     1367     |     428    |        0    |      939 |
| Perl             |        4    |      434     |      49    |       28    |      357 |
| XML              |        3    |      304     |      27    |        0    |      277 |
| Assembly         |        1    |      214     |      20    |       84    |      110 |
| Python           |        1    |       98     |      19    |        2    |       77 |
| Makefile         |        3    |       55     |      14    |        5    |       36 |
| Batch            |        1    |       20     |       1    |        0    |       19 |
| Autoconf         |        1    |       19     |       5    |        4    |       10 |
| Ruby             |        1    |       11     |       1    |        1    |        9 |
|                  |             |              |            |             |          |
| **Total**        |     1345    |   170836     |   15889    |    10873    |   144074 |





JGE/src/main.cpp
#include <pspkernel.h>
#include <pspdisplay.h>
#include <pspdebug.h>
#include <psppower.h>
#include <pspsdk.h>
#include <pspaudiocodec.h>
#include <pspaudio.h>
#include <pspaudiolib.h>
#include <pspmpeg.h>
#include <malloc.h>
#include <string.h>

// Profiling
//#include <pspprof.h>

#include <pspctrl.h>
#include <unistd.h>
#include <stdio.h>
#include <queue>

#include "JGE.h"
#include "JApp.h"
#include "JGameLauncher.h"
#include "JRenderer.h"

#include <wge/log.hpp>

#ifndef JGEApp_Title
#define JGEApp_Title "JGE++"
#endif

#ifdef DEVHOOK
PSP_MODULE_INFO(JGEApp_Title, 0, 1, 1);
PSP_MAIN_THREAD_ATTR(PSP_THREAD_ATTR_USER);
// 256 is not enough for the network to correctly start,
// let's find an appropriate value the day JGE has working network
PSP_HEAP_SIZE_KB(-256);

#else

PSP_MODULE_INFO(JGEApp_Title, 0x1000, 1, 1);
PSP_MAIN_THREAD_ATTR(0);

#endif

int mikModThreadID = -1;
bool done = false;

JApp* game = NULL;
JGE* g_engine = NULL;

u32 gTickFrequency = 1;

//------------------------------------------------------------------------------------------------
// Exit callback
int exit_callback(int arg1, int arg2, void* common) {
    if (g_engine != NULL) g_engine->End();
    sceKernelExitGame();
    return 0;
}

//------------------------------------------------------------------------------------------------
// Power Callback
int power_callback(int unknown, int pwrflags, void* common) {
    if ((pwrflags & (PSP_POWER_CB_POWER_SWITCH | PSP_POWER_CB_STANDBY)) > 0) {
        // suspending
        if (g_engine != NULL) g_engine->Pause();
    } else if ((pwrflags & PSP_POWER_CB_RESUME_COMPLETE) > 0) {
        sceKernelDelayThread(1500000);
        // resume complete
        if (g_engine != NULL) g_engine->Resume();
    }
    return 0;
}

//------------------------------------------------------------------------------------------------
// Callback thread
int CallbackThread(SceSize args, void* argp) {
    int cbid;

#ifdef DEVHOOK
    cbid = sceKernelCreateCallback("Exit Callback", exit_callback, NULL);
    sceKernelRegisterExitCallback(cbid);
#endif
    cbid = sceKernelCreateCallback("Power Callback", power_callback, NULL);
    scePowerRegisterCallback(0, cbid);

    sceKernelSleepThreadCB();

    return 0;
}

// Sets up the callback thread and returns its thread id
int SetupCallbacks(void) {
    int thid = 0;

    thid = sceKernelCreateThread("update_thread", CallbackThread, 0x11, 0xFA0, 0, 0);
    if (thid >= 0) sceKernelStartThread(thid, 0, 0);
    return thid;
}

#ifdef DEVHOOK
// code by sakya, crazyc, samuraiX
// http://forums.ps2dev.org/viewtopic.php?t=9591
PspDebugRegBlock exception_regs;

extern int _ftext;

static const char* codeTxt[32] = {"Interrupt",
                                  "TLB modification",
                                  "TLB load/inst fetch",
                                  "TLB store",
                                  "Address load/inst fetch",
                                  "Address store",
                                  "Bus error (instr)",
                                  "Bus error (data)",
                                  "Syscall",
                                  "Breakpoint",
                                  "Reserved instruction",
                                  "Coprocessor unusable",
                                  "Arithmetic overflow",
                                  "Unknown 14",
                                  "Unknown 15",
                                  "Unknown 16",
                                  "Unknown 17",
                                  "Unknown 18",
                                  "Unknown 19",
                                  "Unknown 20",
                                  "Unknown 21",
                                  "Unknown 22",
                                  "Unknown 23",
                                  "Unknown 24",
                                  "Unknown 25",
                                  "Unknown 26",
                                  "Unknown 27",
                                  "Unknown 28",
                                  "Unknown 29",
                                  "Unknown 31"};

static const unsigned char regName[32][5] = {"zr", "at", "v0", "v1", "a0", "a1", "a2", "a3", "t0", "t1", "t2",
                                             "t3", "t4", "t5", "t6", "t7", "s0", "s1", "s2", "s3", "s4", "s5",
                                             "s6", "s7", "t8", "t9", "k0", "k1", "gp", "sp", "fp", "ra"};

void ExceptionHandler(PspDebugRegBlock* regs) {
    int i;
    SceCtrlData pad;

    pspDebugScreenInit();
    pspDebugScreenSetBackColor(0x00FF0000);
    pspDebugScreenSetTextColor(0xFFFFFFFF);
    pspDebugScreenClear();
    pspDebugScreenPrintf("Your PSP has just crashed!\n");
    pspDebugScreenPrintf("Exception details:\n\n");

    pspDebugScreenPrintf("Exception - %s\n", codeTxt[(regs->cause >> 2) & 31]);
    pspDebugScreenPrintf("EPC       - %08X / %s.text + %08X\n", (int)regs->epc, module_info.modname,
                         (unsigned int)(regs->epc - (int)&_ftext));
    pspDebugScreenPrintf("Cause     - %08X\n", (int)regs->cause);
    pspDebugScreenPrintf("Status    - %08X\n", (int)regs->status);
    pspDebugScreenPrintf("BadVAddr  - %08X\n", (int)regs->badvaddr);
    for (i = 0; i < 32; i += 4)
        pspDebugScreenPrintf("%s:%08X %s:%08X %s:%08X %s:%08X\n", regName[i], (int)regs->r[i], regName[i + 1],
                             (int)regs->r[i + 1], regName[i + 2], (int)regs->r[i + 2], regName[i + 3],
                             (int)regs->r[i + 3]);

    sceKernelDelayThread(1000000);
    pspDebugScreenPrintf("\n\nPress X to dump information on file exception.log and quit");
    pspDebugScreenPrintf("\nPress O to quit");

    for (;;) {
        sceCtrlReadBufferPositive(&pad, 1);
        if (pad.Buttons & PSP_CTRL_CROSS) {
            FILE* log = fopen("exception.log", "w");
            if (log != NULL) {
                char testo[1024] = {0};
                fwrite(testo, 1, strlen(testo), log);
                sprintf(testo, "Exception details:\n\n");
                fwrite(testo, 1, strlen(testo), log);
                sprintf(testo, "Exception - %s\n", codeTxt[(regs->cause >> 2) & 31]);
                fwrite(testo, 1, strlen(testo), log);
                sprintf(testo, "EPC       - %08X / %s.text + %08X\n", (int)regs->epc, module_info.modname,
                        (unsigned int)(regs->epc - (int)&_ftext));
                fwrite(testo, 1, strlen(testo), log);
                sprintf(testo, "Cause     - %08X\n", (int)regs->cause);
                fwrite(testo, 1, strlen(testo), log);
                sprintf(testo, "Status    - %08X\n", (int)regs->status);
                fwrite(testo, 1, strlen(testo), log);
                sprintf(testo, "BadVAddr  - %08X\n", (int)regs->badvaddr);
                fwrite(testo, 1, strlen(testo), log);
                for (i = 0; i < 32; i += 4) {
                    sprintf(testo, "%s:%08X %s:%08X %s:%08X %s:%08X\n", regName[i], (int)regs->r[i], regName[i + 1],
                            (int)regs->r[i + 1], regName[i + 2], (int)regs->r[i + 2], regName[i + 3],
                            (int)regs->r[i + 3]);
                    fwrite(testo, 1, strlen(testo), log);
                }
                fclose(log);
            }
            break;
        } else if (pad.Buttons & PSP_CTRL_CIRCLE)
            break;
        sceKernelDelayThread(100000);
    }
    sceKernelExitGame();
}

void initExceptionHandler() {
    SceKernelLMOption option;
    int args[2], fd, modid;

    memset(&option, 0, sizeof(option));
    option.size = sizeof(option);
    option.mpidtext = PSP_MEMORY_PARTITION_KERNEL;
    option.mpiddata = PSP_MEMORY_PARTITION_KERNEL;
    option.position = 0;
    option.access = 1;

    if ((modid = sceKernelLoadModule("exception.prx", 0, &option)) >= 0) {
        args[0] = (int)ExceptionHandler;
        args[1] = (int)&exception_regs;
        sceKernelStartModule(modid, 8, args, &fd, NULL);
    }
}
#else
//------------------------------------------------------------------------------------------------
// Custom exception handler
void MyExceptionHandler(PspDebugRegBlock* regs) {
    pspDebugScreenInit();

    pspDebugScreenSetBackColor(0x00FF0000);
    pspDebugScreenSetTextColor(0xFFFFFFFF);
    pspDebugScreenClear();

    pspDebugScreenPrintf("I regret to inform you your psp has just crashed\n");
    pspDebugScreenPrintf("Please contact Sony technical support for further information\n\n");
    pspDebugScreenPrintf("Exception Details:\n");
    pspDebugDumpException(regs);
    pspDebugScreenPrintf("\nBlame the 3rd party software, it cannot possibly be our fault!\n");

    sceKernelExitGame();
}

//------------------------------------------------------------------------------------------------
// Sort of hack to install exception handler under USER THREAD
__attribute__((constructor)) void handlerInit() {
    pspKernelSetKernelPC();

    pspSdkInstallNoDeviceCheckPatch();
    pspSdkInstallNoPlainModuleCheckPatch();
    pspSdkInstallKernelLoadModulePatch();

    pspDebugInstallErrorHandler(MyExceptionHandler);
}

#endif

static const struct {
    LocalKeySym keysym;
    JButton keycode;
} gDefaultBindings[] = {
    {PSP_CTRL_START, JGE_BTN_MENU}, {PSP_CTRL_SELECT, JGE_BTN_CTRL},     {PSP_CTRL_RIGHT, JGE_BTN_RIGHT},
    {PSP_CTRL_LEFT, JGE_BTN_LEFT},  {PSP_CTRL_UP, JGE_BTN_UP},           {PSP_CTRL_DOWN, JGE_BTN_DOWN},
    {PSP_CTRL_CIRCLE, JGE_BTN_OK},  {PSP_CTRL_TRIANGLE, JGE_BTN_CANCEL}, {PSP_CTRL_SQUARE, JGE_BTN_PRI},
    {PSP_CTRL_CROSS, JGE_BTN_SEC},  {PSP_CTRL_LTRIGGER, JGE_BTN_PREV},   {PSP_CTRL_RTRIGGER, JGE_BTN_NEXT}};

void JGECreateDefaultBindings() {
    for (signed int i = sizeof(gDefaultBindings) / sizeof(gDefaultBindings[0]) - 1; i >= 0; --i)
        g_engine->BindKey(gDefaultBindings[i].keysym, gDefaultBindings[i].keycode);
}

int JGEGetTime() {
    u64 curr;
    sceRtcGetCurrentTick(&curr);
    return (int)((curr * 1000) / gTickFrequency);
}

static SceCtrlData gCtrlPad;

u8 JGEGetAnalogX() { return gCtrlPad.Lx; }
u8 JGEGetAnalogY() { return gCtrlPad.Ly; }

//------------------------------------------------------------------------------------------------
// The main loop
int main(int argc, char* argv[]) {
    pspDebugScreenInit();

    pspDebugScreenSetBackColor(0x00000000);
    pspDebugScreenSetTextColor(0xFFFFFFFF);
    pspDebugScreenClear();

    pspDebugScreenPrintf("JGE:Loading application...");

    WGE_LOG_TRACE("SetupCallbacks()");
    SetupCallbacks();

#ifdef DEVHOOK
    WGE_LOG_TRACE("initExceptionHandler()");
    initExceptionHandler();
#endif
    g_engine = NULL;

    JGameLauncher* launcher = new JGameLauncher();

    u32 flags = launcher->GetInitFlags();
    if ((flags & JINIT_FLAG_ENABLE3D) != 0) JRenderer::Set3DFlag(true);

    WGE_LOG_TRACE("sceRtcGetTickResolution()");
    gTickFrequency = sceRtcGetTickResolution();
    WGE_LOG_TRACE("JGE::GetInstance()");
    g_engine = JGE::GetInstance();
    g_engine->SetARGV(argc, argv);
    JGECreateDefaultBindings();

    WGE_LOG_TRACE("Create Game");
    game = launcher->GetGameApp();
    game->Create();

    WGE_LOG_TRACE("Run Game");
    g_engine->SetApp(game);
    g_engine->Run();

    game->Destroy();
    delete game;
    game = NULL;

    g_engine->SetApp(NULL);

    done = true;

    delete launcher;

    JGE::Destroy();
    g_engine = NULL;

    // Profiling
    // gprof_cleanup();
    sceKernelExitGame();

    return 0;
}
